src/v/cluster/id_allocator_stm.cc

```diff
@@ -166,19 +166,26 @@ ss::future<> id_allocator_stm::apply(model::record_batch b) {
 
         _processed++;
         if (_processed > _log_capacity) {
-            return _c
-              ->write_snapshot(
-                raft::write_snapshot_cfg(_next_snapshot, iobuf()))
-              .then([this] {
-                  _next_snapshot = _insync_offset;
-                  _processed = 0;
-              });
+            ssx::spawn_with_gate(_gate, [this] { return write_snapshot(); });
         }
     }
 
     return ss::now();
 }
 
+ss::future<> id_allocator_stm::write_snapshot() {
+    if (_is_writing_snapshot) {
+        return ss::now();
+    }
+    _is_writing_snapshot = true;
+    return _c->write_snapshot(raft::write_snapshot_cfg(_next_snapshot, iobuf()))
+      .then([this] {
+          _next_snapshot = _insync_offset;
+          _processed = 0;
+      })
+      .finally([this] { _is_writing_snapshot = false; });
```

Comment:
At this low-level my concern is that the debouncing could fail, but I also don't have the macro view where it might be ok:

Once `_c->write_snapshot(...)` completes there is the scheduling point at then/finally where `_is_writing_snapshot` is true, despite the snapshot having just been written. at that moment a fiber that passed through `apply` and ran `ssx::spawn_with_gate(_gate, [this] { return write_snapshot(); });` would turn into a no-op. the `finally` continuation would run and clear the flag, but the we would have effectively missed the next request to write the snapshot.

the concern there would be the impact on correctness if dropping the request to write the snapshot would be problematic because it could be the case that when it was dropped it was the last time a fiber would pass through apply--that is, no future fiber would come along and kick start things again.

if that is problematic then you could use a semaphore instead of a flag. casually reading through this code though the missed wake up is clear, so we should probably leave a comment if indeed it is safe, so that others don't ruminate on the possibility.

Reply:
The point of calling write_snapshot is to trigger log truncation when `id_allocator` log passes the `_log_capacity` threshold; we won't violate correctness if miss an intent to truncate the log because the conditions which caused this intent won't change and the whole truncation will be retried upon the next request.

Reply:
> intent won't change and the whole truncation will be retried upon the next request.

this all makes sense that the next request will drive a retry. what I'm most interested in is if there isn't a next request. will that be cool?

Reply:
that's fine, there is no much difference between whether we truncate log strict after X records and after X+1. If Redpanda can tolerate X it can definitely tolerate X+1